modules/tensor_mechanics/src/userobjects/CrystalPlasticityStateVarRateComponentVoce.C

```diff
@@ -0,0 +1,271 @@
+                                     "means non-coplanar slip systems in gr_11,22,33= "
+                                     "1.0, 5.0 and 9.0 respectively."
+                                     "latent ahrdening between for gr_12,13 = 2.0 3.0"
+                                     " respectively");
+  params.addClassDescription(
+      "Phenomenological Voce constitutive model state"
+      " variable evolution rate "
+      "component base class.  Override this virtual functions in your class");
+  return params;
+}
+
+CrystalPlasticityStateVarRateComponentVoce::CrystalPlasticityStateVarRateComponentVoce(
+    const InputParameters & parameters)
+  : CrystalPlasticityStateVarRateComponent(parameters),
+    _mat_prop_slip_rate(
+        getMaterialProperty<std::vector<Real>>(parameters.get<std::string>("uo_slip_rate_name"))),
+    _mat_prop_state_var(
+        getMaterialProperty<std::vector<Real>>(parameters.get<std::string>("uo_state_var_name"))),
+    _crystal_lattice_type(getParam<MooseEnum>("crystal_lattice_type")),
+    _groups(getParam<std::vector<unsigned int>>("groups")),
+    _h0_group_values(getParam<std::vector<Real>>("h0_group_values")),
+    _tau0_group_values(getParam<std::vector<Real>>("tau0_group_values")),
+    _tauSat_group_values(getParam<std::vector<Real>>("tauSat_group_values")),
+    _hardeningExponent_group_values(getParam<std::vector<Real>>("hardeningExponent_group_values")),
+    _selfHardening_group_values(getParam<std::vector<Real>>("selfHardening_group_values")),
+    _coplanarHardening_group_values(getParam<std::vector<Real>>("coplanarHardening_group_values")),
+    _GroupGroup_Hardening_group_values(
+        getParam<std::vector<Real>>("GroupGroup_Hardening_group_values"))
+{
+  // perform input checks and initialize usefull variables
+  _n_groups = _groups.size() - 1;
+  checkHardeningParametersSize();
+  initSlipSystem_PlaneID(_slipSystem_PlaneID);
+  initSlipSystem_GroupID(_slipSystem_GroupID);
+}
+
+void
+CrystalPlasticityStateVarRateComponentVoce::checkHardeningParametersSize() const
+{
+  // check that at least one group exists
+  if (_n_groups <= 0)
+    mooseError("CrystalPlasticityStateVarRateComponentVoce: Error in reading hardening     "
+               "parameters values "
+               "Specify input in .i file ");
+
+  // check the size of all the variables
+  bool check_var_size = true;
+  check_var_size &= _h0_group_values.size() == _n_groups;
+  check_var_size &= _tau0_group_values.size() == _n_groups;
+  check_var_size &= _tauSat_group_values.size() == _n_groups;
+  check_var_size &= _hardeningExponent_group_values.size() == _n_groups;
+  check_var_size &= _selfHardening_group_values.size() == _n_groups;
+  check_var_size &= _coplanarHardening_group_values.size() == _n_groups;
+  check_var_size &= _h0_group_values.size() == _n_groups;
+  check_var_size &= _GroupGroup_Hardening_group_values.size() == _n_groups * _n_groups;
```

Comment:
I'd write this as

```
  const bool check_var_size =
      _h0_group_values.size() == _n_groups && _tau0_group_values.size() == _n_groups &&
      _tauSat_group_values.size() == _n_groups &&
      _hardeningExponent_group_values.size() == _n_groups &&
      _selfHardening_group_values.size() == _n_groups &&
      _coplanarHardening_group_values.size() == _n_groups && _h0_group_values.size() == _n_groups &&
      _GroupGroup_Hardening_group_values.size() == _n_groups * _n_groups;
```

but for a gold star you should really have a `paramError` for each parameter that fails the check. Otherwise the user will have to go through 8 input parameter vectors, count the elements in each one, and find the one that has the wron size (without any indication what the _right_ size is).

Reply:
Now each parameters has its own test.